[BUG] node-canvas error

### Bug Description

bacause of node-canvas made the dev down, I temporally removed it by #3978

```shell
2024-08-21T15:01:19.799808009Z: [ERROR]  Error: libuuid.so.1: cannot open shared object file: No such file or directory
2024-08-21T15:01:19.799811309Z: [ERROR]      at Module._extensions..node (node:internal/modules/cjs/loader:1454:18)
2024-08-21T15:01:19.799814609Z: [ERROR]      at Module.load (node:internal/modules/cjs/loader:1208:32)
2024-08-21T15:01:19.799818009Z: [ERROR]      at Module._load (node:internal/modules/cjs/loader:1024:12)
2024-08-21T15:01:19.799821109Z: [ERROR]      at Module.require (node:internal/modules/cjs/loader:1233:19)
2024-08-21T15:01:19.799824209Z: [ERROR]      at require (node:internal/modules/helpers:179:18)
2024-08-21T15:01:19.799827309Z: [ERROR]      at Object.<anonymous> (/geohub/node_modules/.pnpm/canvas@2.11.2/node_modules/canvas/lib/bindings.js:3:18)
2024-08-21T15:01:19.799831209Z: [ERROR]      at Module._compile (node:internal/modules/cjs/loader:1358:14)
2024-08-21T15:01:19.799834409Z: [ERROR]      at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
2024-08-21T15:01:19.799837509Z: [ERROR]      at Module.load (node:internal/modules/cjs/loader:1208:32)
2024-08-21T15:01:19.799840710Z: [ERROR]      at Module._load (node:internal/modules/cjs/loader:1024:12) {
2024-08-21T15:01:19.799843910Z: [ERROR]    code: 'ERR_DLOPEN_FAILED'
```

### Bug Reproduction

revert changes of #3978 and deploy to appservice

### Bug Impact

_No response_

### Bug Screenshots

_No response_

### Bug References

_No response_

### Bug System Information

- OS: [e.g. Windows 10]
- Browser: [e.g. Google Chrome, Firefox]

--- a/.changeset/healthy-lemons-rest.md
+++ b/.changeset/healthy-lemons-rest.md
@@ -1,0 +1,5 @@
+---
+"geohub": patch
+---
+
+fix: use pngjs to crop sprite image instead of using image-clipper with node-canvas

--- a/sites/geohub/package.json
+++ b/sites/geohub/package.json
@@ -58,6 +58,7 @@
 		"@types/papaparse": "^5.3.14",
 		"@types/pbf": "^3.0.5",
 		"@types/pg": "^8.11.6",
+		"@types/pngjs": "^6.0.5",
 		"@types/uuid": "^10.0.0",
 		"@typescript-eslint/eslint-plugin": "8.1.0",
 		"@typescript-eslint/parser": "^8.0.0",
@@ -146,7 +147,8 @@
 		"crypto-js": "^4.2.0",
 		"jwt-decode": "^4.0.0",
 		"pbf": "^4.0.1",
-		"pg": "^8.12.0"
+		"pg": "^8.12.0",
+		"pngjs": "^7.0.0"
 	},
 	"type": "module"
 }

--- a/sites/geohub/src/lib/server/SvgLegendCreator.ts
+++ b/sites/geohub/src/lib/server/SvgLegendCreator.ts
@@ -1,3 +1,4 @@
+import { getDecimalPlaces } from '$lib/helper';
 import chroma from 'chroma-js';
 import { hexToCSSFilter } from 'hex-to-css-filter';
 
@@ -50,6 +51,18 @@ export class SvgLegendCreator {
 		colors: [number, number, number, number][],
 		options?: SvgLegendCreatorOptions
 	) {
+		let minDecimalPlaces = 0;
+		if (options?.min && typeof options.min === 'string') {
+			options.min = parseFloat(options.min);
+			minDecimalPlaces = getDecimalPlaces(options.min);
+			if (minDecimalPlaces > 2) minDecimalPlaces = 2;
+		}
+		let maxDecimalPlaces = 0;
+		if (options?.max && typeof options.max === 'string') {
+			options.max = parseFloat(options.max);
+			maxDecimalPlaces = getDecimalPlaces(options.max);
+			if (maxDecimalPlaces > 2) maxDecimalPlaces = 2;
+		}
 		const contents = `
         <defs>
             <linearGradient id='grad1' x1='0%' y1='0%' x2='100%' y2='0%'>
@@ -63,9 +76,9 @@ export class SvgLegendCreator {
             </linearGradient>
         </defs>
         ${options?.unit ? `<text x='0' y='15' font-family='${this.fontFamily}' font-size='${this.fontSize}' fill='#000000'>${options?.unit}</text>` : ''}
-        <rect y='20' width='100%' height='28' fill='url(#grad1)' />
-        ${options?.min ? `<text x='0' y='${options?.unit ? 65 : 45}' font-family='${this.fontFamily}' font-size='${this.fontSize}' fill='#000000'>${options?.min.toFixed(0)}</text>` : ''}
-        ${options?.max ? `<text x='100%' y='${options?.unit ? 65 : 45}' font-family='${this.fontFamily}' font-size='${this.fontSize}' fill='#000000' text-anchor='end'>${options?.max.toFixed(0)}</text>` : ''}
+        <rect y='${options?.unit ? 20 : 0}' width='100%' height='28' fill='url(#grad1)' />
+        ${options?.min ? `<text x='0' y='${options?.unit ? 65 : 45}' font-family='${this.fontFamily}' font-size='${this.fontSize}' fill='#000000'>${options?.min.toFixed(minDecimalPlaces)}</text>` : ''}
+        ${options?.max ? `<text x='100%' y='${options?.unit ? 65 : 45}' font-family='${this.fontFamily}' font-size='${this.fontSize}' fill='#000000' text-anchor='end'>${options?.max.toFixed(maxDecimalPlaces)}</text>` : ''}
 `;
 
 		const height = options?.unit ? 70 : 50;

--- a/sites/geohub/src/lib/server/helpers/clipSpriteServer.ts
+++ b/sites/geohub/src/lib/server/helpers/clipSpriteServer.ts
@@ -1,28 +1,40 @@
-import Clipper from 'image-clipper';
-import type { SpriteIcon, SpriteImage } from '../../types';
-import canvas from 'canvas';
+import { PNG } from 'pngjs';
+import type { SpriteIcon, SpriteImage } from '$lib/types';
 
-/**
- * Clip an image by id from sprite file image (server version)
- * @param url sprite image URL
- * @param id Sprite ID
- * @param icon sprite image data from json
- * @returns
- */
-export const clipSpriteServer = (
+export const clipSpriteServer = async (
 	url: string,
 	id: string,
 	icon: SpriteIcon
 ): Promise<SpriteImage> => {
-	return new Promise((resolve) => {
-		Clipper.configure({ canvas: canvas });
-		Clipper(url, function () {
-			this.crop(icon.x, icon.y, icon.width, icon.height).toDataURL(function (dataUrl: string) {
-				resolve({
-					src: dataUrl,
-					alt: id
-				});
-			});
-		});
-	});
+	const res = await fetch(url);
+	if (!res.ok) {
+		throw new Error(`Failed to fetch image: ${res.statusText}`);
+	}
+
+	const arrayBuffer = await res.arrayBuffer();
+	const buffer = Buffer.from(arrayBuffer);
+
+	const png = PNG.sync.read(buffer);
+	const cropped = new PNG({ width: icon.width, height: icon.height });
+
+	for (let y = 0; y < icon.height; y++) {
+		for (let x = 0; x < icon.width; x++) {
+			const idx = ((icon.y + y) * png.width + (icon.x + x)) << 2;
+			const croppedIdx = (y * icon.width + x) << 2;
+
+			cropped.data[croppedIdx] = png.data[idx];
+			cropped.data[croppedIdx + 1] = png.data[idx + 1];
+			cropped.data[croppedIdx + 2] = png.data[idx + 2];
+			cropped.data[croppedIdx + 3] = png.data[idx + 3];
+		}
+	}
+
+	const bufferOutput = PNG.sync.write(cropped);
+	const base64 = bufferOutput.toString('base64');
+	const dataUrl = `data:image/png;base64,${base64}`;
+
+	return {
+		src: dataUrl,
+		alt: id
+	};
 };

--- a/sites/geohub/src/lib/server/helpers/generateLegendFromStyle.ts
+++ b/sites/geohub/src/lib/server/helpers/generateLegendFromStyle.ts
@@ -1,6 +1,6 @@
 import {
 	convertFunctionToExpression,
-	// fetchUrl,
+	fetchUrl,
 	getActiveBandIndex,
 	getDecimalPlaces,
 	isRgbRaster
@@ -13,7 +13,7 @@ import type {
 	DashboardMapStyle,
 	Layer,
 	RasterTileMetadata,
-	// SpriteIcon,
+	SpriteIcon,
 	VectorLayerSpecification,
 	VectorTileMetadata
 } from '$lib/types';
@@ -22,8 +22,8 @@ import type {
 	RasterSourceSpecification,
 	SpriteSpecification
 } from 'maplibre-gl';
-// import { clipSpriteServer } from './clipSpriteServer';
 import { layerTypes } from '@undp-data/svelte-maplibre-storymap';
+import { clipSpriteServer } from './clipSpriteServer';
 
 /**
  * LegendLayer interface to contain layer legend information
@@ -262,25 +262,23 @@ const getVectorPropertyNames = async (
 		colorProp = 'heatmap-color';
 	} else if (layer.type === 'symbol') {
 		colorProp = 'icon-color';
-		console.log(sprite);
-		// const iconName = layer.layout ? (layer.layout['icon-image'] as string) : undefined;
-		// if (iconName) {
-		// 	if (typeof sprite === 'string') {
-		// 		const spriteBase = sprite.replace('/sprite/sprite', '/sprite-non-sdf/sprite');
-		// 		// const spriteBase = sprite;
-		// 		const data = `${spriteBase}@2x.png`;
-		// 		const spriteJson: { [key: string]: SpriteIcon } = (await fetchUrl(
-		// 			`${spriteBase}@2x.json`
-		// 		)) as unknown as { [key: string]: SpriteIcon };
-		// 		const spriteImage = spriteJson[iconName];
-		// 		if (spriteImage) {
-		// 			const image = await clipSpriteServer(data, iconName, spriteImage);
-		// 			shape = `
-		// 			 <image x='0' y='0' width='{size}' height='{size}' xlink:href='${image.src}' style='{style}' />
-		// 			`;
-		// 		}
-		// 	}
-		// }
+		const iconName = layer.layout ? (layer.layout['icon-image'] as string) : undefined;
+		if (iconName) {
+			if (typeof sprite === 'string') {
+				const spriteBase = sprite.replace('/sprite/sprite', '/sprite-non-sdf/sprite');
+				const data = `${spriteBase}@2x.png`;
+				const spriteJson: { [key: string]: SpriteIcon } = (await fetchUrl(
+					`${spriteBase}@2x.json`
+				)) as unknown as { [key: string]: SpriteIcon };
+				const spriteImage = spriteJson[iconName];
+				if (spriteImage) {
+					const image = await clipSpriteServer(data, iconName, spriteImage);
+					shape = `
+					 <image x='0' y='0' width='{size}' height='{size}' xlink:href='${image.src}' style='{style}' />
+					`;
+				}
+			}
+		}
 	}
 
 	return {

--- a/sites/geohub/src/lib/server/helpers/generateLegendFromStyle.ts
+++ b/sites/geohub/src/lib/server/helpers/generateLegendFromStyle.ts
@@ -7,7 +7,7 @@ import {
 } from '$lib/helper';
 import chroma from 'chroma-js';
 import { hexToCSSFilter } from 'hex-to-css-filter';
-import { SvgLegendCreator, type SvgLegendCreatorOptions } from '../SvgLegendCreator';
+import { SvgLegendCreator, type SvgLegendCreatorOptions } from '$lib/server/SvgLegendCreator';
 import type {
 	BandMetadata,
 	DashboardMapStyle,
@@ -122,14 +122,20 @@ const getRasterLayerLegend = async (geohubLayer: Layer, source: RasterSourceSpec
 	const url = new URL(sourceUrl);
 
 	const algorithmId = url.searchParams.get('algorithm');
+	const colormap_name = url.searchParams.get('colormap_name');
+	const colormapString = url.searchParams.get('colormap');
 
 	let legend = '';
 	const creatorOption: SvgLegendCreatorOptions = {};
 	let colors: [number, number, number, number][] = [];
 	let values: string[] | number[][] = [];
-	if (!isRgbTile && !algorithmId) {
-		const colormap_name = url.searchParams.get('colormap_name');
-		const colormapString = url.searchParams.get('colormap');
+
+	if (
+		// non-true color raster without algorithm
+		(!isRgbTile && !algorithmId) ||
+		// raster with algorithm and customised colormap/colors
+		(algorithmId && (colormap_name || colormapString))
+	) {
 		let colormap: number[][][] | { [key: string]: number[] } = [];
 		if (colormapString) {
 			if (Array.isArray(colormapString)) {
@@ -195,7 +201,7 @@ const getRasterLayerLegend = async (geohubLayer: Layer, source: RasterSourceSpec
 			}
 		}
 	} else {
-		// RGB legend or algorithm
+		// RGB legend or algorithm without colormap/colors
 
 		const links = geohubLayer.dataset?.properties.links;
 		const titilerBaseUrl = links?.find((l) => l.rel === 'cog')?.href;

